openedx/core/djangoapps/waffle_utils/__init__.py

```diff
@@ -0,0 +1,255 @@
+        with waffle_override_switch(namespaced_switch_name, active):
+            log.info(u"%sSwitch '%s' set to %s in model.", self.log_prefix, namespaced_switch_name, active)
+            yield
+
+    @property
+    def _cached_switches(self):
+        """
+        Returns a dictionary of all namespaced switches in the request cache.
+        """
+        return self._get_namespace_request_cache().setdefault('switches', {})
+
+
+class WaffleFlagNamespace(WaffleNamespace):
+    """
+    Provides a request cached namespace for waffle flags.
+    """
+    __metaclass__ = ABCMeta
+
+    @property
+    def _cached_flags(self):
+        """
+        Returns a dictionary of all namespaced flags in the request cache.
+        """
+        return self._get_namespace_request_cache().setdefault('flags', {})
+
+    def is_enabled(self, flag_name, check_before_waffle_callback=None):
+        """
+        Returns and caches whether the given flag is enabled.
+
+        If the flag value is already cached in the request, it is returned.
+        If check_before_waffle_callback is supplied, it is called before
+            checking waffle.
+        If check_before_waffle_callback returns None, or if it is not supplied,
+            then waffle is used to check the flag.
+
+        Arguments:
+            flag_name (String): The name of the flag to check.
+            check_before_waffle_callback (function): (Optional) A function that
+                will be checked before continuing on to waffle. If
+                check_before_waffle_callback(namespaced_flag_name) returns True
+                or False, it is cached and returned.  If it returns None, then
+                waffle is used.
+
+        """
+        # validate arguments
+        namespaced_flag_name = self._namespaced_name(flag_name)
+        value = self._cached_flags.get(namespaced_flag_name)
+
+        if value is None:
+            if check_before_waffle_callback:
+                value = check_before_waffle_callback(namespaced_flag_name)
+
+            if value is None:
+                value = flag_is_active(get_request(), namespaced_flag_name)
+
+            self._cached_flags[namespaced_flag_name] = value
+        return value
+
+
+class WaffleFlag(object):
+    """
+    Represents a single waffle flag, using a cached waffle namespace.
+    """
+
+    def __init__(self, waffle_namespace, flag_name):
+        """
+        Initializes the waffle flag instance.
+
+        Arguments:
+            waffle_namespace (WaffleFlagNamespace): Provides a cached namespace
+                for this flag.
+            flag_name (String): The name of the flag (without namespacing).
+
+        """
+        self.waffle_namespace = waffle_namespace
+        self.flag_name = flag_name
+
+    def is_enabled(self):
+        """
+        Returns whether or not the flag is enabled.
+        """
+        return self.waffle_namespace.is_enabled(self.flag_name)
+
+
+class CourseOverrideWaffleFlag(WaffleFlag):
```

Comment:
Nit: I don't love this name. It isn't a flag that overrides courses, it is a flag that supports per-course overrides. I think ``CourseWaffleFlag`` might be clear enough. If that isn't specific enough, then maybe ``CourseAwareWaffleFlag``?